File: collect_app/src/main/java/org/odk/collect/android/configure/qr/QRCodeActivityResultDelegate.java
package org.odk.collect.android.configure.qr;

import static android.app.Activity.RESULT_OK;
import static org.odk.collect.android.activities.ActivityUtils.startActivityAndCloseAllOthers;
import static org.odk.collect.android.configure.qr.QRCodeMenuDelegate.SELECT_PHOTO;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.widget.Toast;

import org.odk.collect.analytics.Analytics;
import org.odk.collect.android.R;
import org.odk.collect.android.activities.MainMenuActivity;
import org.odk.collect.android.analytics.AnalyticsEvents;
import org.odk.collect.android.utilities.ActivityResultDelegate;
import org.odk.collect.projects.Project;
import org.odk.collect.settings.ODKAppSettingsImporter;

import java.io.FileNotFoundException;
import java.io.InputStream;

public class QRCodeActivityResultDelegate implements ActivityResultDelegate {

    private final Activity activity;
    private final ODKAppSettingsImporter settingsImporter;
    private final QRCodeDecoder qrCodeDecoder;
    private final Project.Saved project;

    public QRCodeActivityResultDelegate(Activity activity, ODKAppSettingsImporter settingsImporter,
                                        QRCodeDecoder qrCodeDecoder, Project.Saved project) {
        this.activity = activity;
        this.settingsImporter = settingsImporter;
        this.qrCodeDecoder = qrCodeDecoder;
        this.project = project;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == SELECT_PHOTO && resultCode == RESULT_OK && data != null) {
            final Uri imageUri = data.getData();
            if (imageUri != null) {
                final InputStream imageStream;

                try {
                    imageStream = activity.getContentResolver().openInputStream(imageUri);
                } catch (FileNotFoundException e) {
                    // Not sure how this could happen? If you work it out: write a test!
                    return;
                }

                try {
                    String response = qrCodeDecoder.decode(imageStream);
                    if (response != null) {
                        if (settingsImporter.fromJSON(response, project)) {
                            Analytics.log(AnalyticsEvents.RECONFIGURE_PROJECT);
                            showToast(R.string.successfully_imported_settings);
                            startActivityAndCloseAllOthers(activity, MainMenuActivity.class);
                        } else {
                            showToast(R.string.invalid_qrcode);
                        }
                    }

                } catch (QRCodeDecoder.InvalidException e) {
                    showToast(R.string.invalid_qrcode);
                } catch (QRCodeDecoder.NotFoundException e) {
                    showToast(R.string.qr_code_not_found);
                }
            }
        }
    }

    private void showToast(int string) {
        Toast.makeText(activity, activity.getString(string), Toast.LENGTH_LONG).show();
    }
}

File: collect_app/src/main/java/org/odk/collect/android/configure/qr/QRCodeActivityResultDelegate.kt
package org.odk.collect.android.configure.qr

import android.app.Activity
import android.content.Intent
import android.widget.Toast
import org.odk.collect.analytics.Analytics.Companion.log
import org.odk.collect.android.R
import org.odk.collect.android.activities.ActivityUtils
import org.odk.collect.android.analytics.AnalyticsEvents
import org.odk.collect.android.mainmenu.MainMenuActivity
import org.odk.collect.projects.Project.Saved
import org.odk.collect.qrcode.QRCodeDecoder
import org.odk.collect.settings.ODKAppSettingsImporter
import org.odk.collect.settings.importing.SettingsImportingResult
import java.io.FileNotFoundException
import java.io.InputStream

class QRCodeActivityResultDelegate(
    private val activity: Activity,
    private val settingsImporter: ODKAppSettingsImporter,
    private val qrCodeDecoder: QRCodeDecoder,
    private val project: Saved
) {
    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == QRCodeMenuDelegate.SELECT_PHOTO && resultCode == Activity.RESULT_OK && data != null) {
            val imageUri = data.data
            if (imageUri != null) {
                val imageStream: InputStream? = try {
                    activity.contentResolver.openInputStream(imageUri)
                } catch (e: FileNotFoundException) {
                    // Not sure how this could happen? If you work it out: write a test!
                    return
                }
                try {
                    val response = qrCodeDecoder.decode(imageStream)

                    when (settingsImporter.fromJSON(response, project)) {
                        SettingsImportingResult.SUCCESS -> {
                            log(AnalyticsEvents.RECONFIGURE_PROJECT)
                            showToast(R.string.successfully_imported_settings)
                            ActivityUtils.startActivityAndCloseAllOthers(
                                activity,
                                MainMenuActivity::class.java
                            )
                        }
                        SettingsImportingResult.INVALID_SETTINGS -> showToast(R.string.invalid_qrcode)
                        SettingsImportingResult.GD_PROJECT -> showToast(R.string.settings_with_gd_protocol)
                    }
                } catch (e: QRCodeDecoder.QRCodeInvalidException) {
                    showToast(R.string.invalid_qrcode)
                } catch (e: QRCodeDecoder.QRCodeNotFoundException) {
                    showToast(R.string.qr_code_not_found)
                }
            }
        }
    }

    private fun showToast(string: Int) {
        Toast.makeText(activity, activity.getString(string), Toast.LENGTH_LONG).show()
    }
}
